classes/query.php

```diff
@@ -1608,6 +1628,15 @@ public function process_row($row, $models, &$result)
 					}
 				}
 
+				// If this is a relation without records, set it to an empty array and continue
+				// Also check for [] because a previous loop iteration may have already set it up
+				if ($is_relation_without_records and (! isset($target[$current]) or $target[$current] === []) )
+				{
+					$target[$current] = [];
```

Comment:
Also, the missing relation could be a "one" or a "many", so I wonder if that should be checked here, because singular relations shouldn't be initialized with an empty array, but with null?

Reply:
I through the same, but since in the model the check for relation is only done for array(), I wouldn't modify too much this part of the code to avoid causing unexpected regressions.

Reply:
Not sure that is true.

In https://github.com/fuel/orm/blob/1.9/develop/classes/model.php#L1220 it will check if related data exists by means of an `array_key_exists` (which will be true after this code change),  and it then procedes to return the value.

Now, application code will for a HasOne relation expect the value to be either `null` (no related object) or an instance of `\Orm\Model`. But not an array. So although the ORM itself won't be too bothered with it, it is possible that user code will fail.

So I think it's probably better to replace it by
````
$target[$current] = $model['singular'] ? null : array();
````

Reply:
I wrote it that way in the first place, but the line 2077 of model.php, which is : 
`elseif (array_key_exists($property, static::relations()) and is_array($value))`
Check that the value is an array to enter the code instruction block that handle the relation part. So, your proposal would require to change that line for : 
`elseif (array_key_exists($property, static::relations()) and (is_array($value)) or $value === null)`
I didn't do it to avoid causing regressions, even if there shouldn't be any issue. Tell me what you prefer.

Reply:
The `$value` used there comes from the argument `$values` of `from_array()`, so not related to this issue.

For converting array data into model objects, it should always be an array, either an array of values (for singlar objects) or an array of arrays (for non-singular objects).

So it's say your first hunch was correct ;-).

Reply:
p.s. where in `query.php` is `from_array()` used? Can't seem to find it.

Reply:
Trace : 
**Query.php -> get() -> l.1469** 
```
$objects[$key] = $this->model::forge($record, false, $this->view ? $this->view['_name'] : null, static::$caching);
```
**Model.php -> forge() -> l.145** 
```
return new static($data, $new, $view, $cache); // Call constructor
```
**Model.php -> __construct() -> l.923** 
```
$this->from_array($data);
```
**Model.php -> from_array() -> l.2077** 
```
elseif (array_key_exists($property, static::relations()) and is_array($value))
```
My first hunch was invalidated by tests, forcing me to trace back why singular relations set to null instead of [] weren't defined. The last line should be modified as given in my last comment for it to work.

Reply:
Hmm... Bloody difficult if all you have is a webpage with code, and no system to actually test on... :disappointed: 

I get it now. The result of the hydration is fed into `Model::forge()` which in turn calls `from_array()` to create the model objects. Which does indeed requires arrays, the correct initialisation is done on L#2082.

Which means no more points of discussion, this can be merged, right?

Reply:
Right. :-)